From aba82dfd2fc752744f430ab4b6d15723bc748663 Mon Sep 17 00:00:00 2001
From: K Prateek Nayak <kprateek.nayak@amd.com>
Date: Fri, 26 Apr 2024 12:46:11 +0200
Subject: [PATCH 11/12] sched/eevdf: Limit preemption a little more

XXX changelog goes here

TL;DR, adding tasks to the right of the tree moves avg_vruntime right,
which possibly makes curr non-eligible and causes insta-preemption
irrespective of it only having just landed on the CPU.

Signed-off-by: Peter Zijlstra (Intel) <peterz@infradead.org>
---
 kernel/sched/fair.c     |  4 ++--
 kernel/sched/features.h | 11 +++++++----
 2 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index e75775542..0fe98b970 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -867,7 +867,7 @@ static inline bool pick_curr(struct cfs_rq *cfs_rq,
 	/*
 	 * Nothing to preserve...
 	 */
-	if (!curr || !sched_feat(RUN_TO_PARITY))
+	if (!curr || !sched_feat(RESPECT_SLICE))
 		return false;
 
 	/*
@@ -875,7 +875,7 @@ static inline bool pick_curr(struct cfs_rq *cfs_rq,
 	 * is consumed. Note: placement of positive lag can push V left and render
 	 * @curr instantly ineligible irrespective the time on-cpu.
 	 */
-	if (!entity_eligible(cfs_rq, curr))
+	if (sched_feat(RUN_TO_PARITY) && !entity_eligible(cfs_rq, curr))
 		return false;
 
 	/*
diff --git a/kernel/sched/features.h b/kernel/sched/features.h
index 64ce99cf0..2285dc302 100644
--- a/kernel/sched/features.h
+++ b/kernel/sched/features.h
@@ -10,12 +10,15 @@ SCHED_FEAT(PLACE_LAG, true)
  */
 SCHED_FEAT(PLACE_DEADLINE_INITIAL, true)
 /*
- * Inhibit (wakeup) preemption until the current task has either matched the
- * 0-lag point or until is has exhausted it's slice.
+ * Inhibit (wakeup) preemption until the current task has exhausted its slice.
  */
-SCHED_FEAT(RUN_TO_PARITY, true)
+SCHED_FEAT(RESPECT_SLICE, true)
 /*
- * Allow tasks with a shorter slice to disregard RUN_TO_PARITY
+ * Relax RESPECT_SLICE to allow preemption once current has reached 0-lag.
+ */
+SCHED_FEAT(RUN_TO_PARITY, false)
+/*
+ * Allow tasks with a shorter slice to disregard RESPECT_SLICE
  */
 SCHED_FEAT(PREEMPT_SHORT, true)
 
-- 
2.44.0.325.g11c821f2f2

